### DISCUSSION TITLE
Cypress Custom Command does not run in test-body

### DISCUSSION THREAD

**Discussion by Shaegi** [cypress-io/cypress#25338] [Questions and Help]
I'm trying to write a plugin which compares DOM-Snapshots. To do so I'm writing the element into a file in reading it on the next run.
The functionallity is working fine, however the custom-command does not execute in the test-body, but fails/succeeds in the 'after-all'-Hook.
After crawling through github/stackoverflow issues, I couldn't find a solution for my problem.

The plugin code is the following:
````typescript
Cypress.Commands.add<'domSnapShot', JQuery<any>>(
  'domSnapShot',
  {
    prevSubject: true
  },
  (subject, options = {}) => {
    return cy
      .task(
        TASKS.getSnapshotPath,
        {
          title: Cypress.currentTest.titlePath.join(' '),
          specPath: Cypress.spec.relative
        },
        { log: false }
      )
      .then((res: { path: String }) => {
        return cy
          .task(
            TASKS.compareSnapshots,
            {
              pathToFile: res.path,
              newElement: subject.prop('outerHTML')
            },
            {
              log: true
            }
          )
          .then(
            (res: {
              oldContent?: string
              newContent?: string
            }) => {
               expect(res.oldContent).to.equal(res.newContent)
              return {
                ...res,
                subject
              }
            }
          )
      })
  }
)

````

The test code is the fowlloing: 

````typescript
describe('<ScrollContainer />', () => {
  it('matches Snapshot', async () => {
// [...]
    cy.get('body').domSnapShot()
  })
})
````

In the runner im getting this output:
![image](https://user-images.githubusercontent.com/33374188/210367929-ee3dc0af-04d9-4afb-ad22-3851e2582db6.png)

if I have multiple tests with my custom command, only the last will be executed. 
I tried to have a custom cy.wait with alias but it didn't help. 
I hope you can help me! Thanks!

**Comment by portyu9** [comment]
Hi @Shaegi 

The first thing I would change is removing `async` from the test.

Cypress commands are queued commands, **not Promises**, and should not be mixed with an `async` Mocha test like this:

```ts
it('matches Snapshot', async () => {
  cy.get('body').domSnapShot()
})
```

Use:

```ts
it('matches Snapshot', () => {
  return cy.get('body').domSnapShot()
})
```

Your custom command already returns the Cypress chain, which is the important part:

```ts
return cy
  .task(...)
  .then(...)
```

For an element child command I would also make the subject requirement explicit:

```ts
Cypress.Commands.add(
  'domSnapShot',
  { prevSubject: 'element' },
  (subject, options = {}) => {
    return cy
      .task(/* ... */)
      .then(/* ... */)
  }
)
```

A fixed `cy.wait()` should not be necessary here. Cypress will wait for the returned command chain.

If removing `async` still produces the same behavior, the runner output/reproduction would be needed to diagnose further, because the custom command implementation shown here does return its asynchronous Cypress work correctly.

Please mark this answer as accepted if it helped, thank you.

**Comment by portyu9** [comment]
Hi @Shaegi 

`cy.visit({ headers })` only adds those headers to the **initial document request**. It does not automatically add them to the subsequent requests made by your application.

For requests made by the browser, register an intercept **before** visiting the page and modify the outgoing header:

```js
cy.intercept(
  { url: 'https://your-api.example.com/**', middleware: true },
  (req) => {
    req.headers['accept-language'] = 'fr'
  }
)

cy.visit('/')
```

You can broaden or narrow the URL matcher depending on which requests need the language header.

Changing:

```js
navigator.language
```

does not change the HTTP `Accept-Language` header, so that approach will not solve this.

There is one important distinction in your description: if the browser calls **Server A**, and Server A then makes its own server-to-server request to **Server B**, Cypress cannot modify that second request. `cy.intercept()` only sees traffic passing through Cypress's browser proxy.

In that case the correct flow is:

```text
browser -- Accept-Language: fr --> Server A
                                  |
                                  +-- Server A must forward/set it --> Server B
```

So Cypress can provide `Accept-Language` on the browser request, but your backend must propagate it if a downstream server also needs it.

Please mark this answer as accepted if it helped, thank you.
